### PR TITLE
Weighted pMDS - Feature/exe-1782

### PR DESCRIPTION
**Pull request by ludvigla**
## Description

Adds additional arguments to `layout_pmds` to use edge weights and noise.

The new option `use_weights` accepts a numpy array with any non-negative edge weights with one element per edge. Alternatively, one can set it to 'prob_dist' which activates the use of edge weights computed from transition probabilities. The default is None, in which case no edge weights are used. 

### Brief explanation of edge weights

Weights are calculated for edges in the graph from the bidirectional transition probability of a random walker. For an undirected edge $e_{i,j}$, we define the bidirectional transition probability $P(e_{u,v})$ as the probability of transitioning from $u$ to $v$ and back to $u$ in $k$ steps. Note that it doesn't matter what node the random walker starts from since it has to go in both directions. In this computation, we also allow self-loops in the graph to slow the transition of the random walker.

Then, we assume that there is a relationship between the bidirectional transition probability and an exponential function which decays with distance:

$P(e_{u,v}) \propto e^{-d(u, v)}$

If the relationship holds, we can solve for $d(u, v)$ to estimate weights that should be proportional to distance: 

$w_{u,v} = -\log(P(e_{u,v}))$

Examples demonstrating this weighting scheme is available in `layout_with_weighted_pmds` in pixelatorR-internal. The strategy is entirely based on empirical observations but does seem to improve the variance explained when computing pMDS layouts substantially.

Fixes: exe-1782

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce it when relevant.

## PR checklist:

- [x] This comment contains a description of changes (with reason).
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and documentation and corrected any misspellings
- [x] I have documented any significant changes to the code in [CHANGELOG.md](../CHANGELOG.md)

### CODE REVIEW

**Comment by ambarrio** [comment]
Check tests not passing.

**Review by ptajvar — APPROVED** [review]
Looks great to me!

**Comment by ludvigla** [comment]
@ambarrio , now I have added wpmds_3d as the default option in `pmds_layout` and it should be possible to run it with the pre-compute layout cli option.

A few things to think about here:

- I have set the default number of pivot points in pMDS to 200 instead of 50. I think 50 might be a bit low for larger graphs. 
- It is possible that some graphs might have fewer than 200 pivot points and in that case the layout computation will fail. Maybe we should resort to using all available nodes as pivot points if the total number of nodes is lower than 200.

**Comment by ambarrio** [comment]
I can review if it is ready to go.

1- good
2- ok, but I think we run layout as part of the nf-core/pixelator and we never come down to 200 nodes in annotated cells. But as a safeguard when it is run under an API call or from CLI in the unannotated PXL file, makes sense.

**Comment by ludvigla** [comment]
For point 2, I'm just not sure where that check should be placed.

There are still some tests failing, and I cannot figure out why. There's a mismatch in the generated figures when running the 2d_plot tests. My suspicion is that it's related to the `setup_basic_pixel_dataset`.

**Comment by ambarrio** [comment]
Probably just here: https://github.com/PixelgenTechnologies/pixelator/blob/482a2c7811de8a4ecc21cca521f9ce20dda3fbb0/src/pixelator/graph/backends/implementations/_networkx.py#L768

where we can write: `pivots = len(g.nodes)`

and then, it they become very few (i.e. <50), they will raise here: https://github.com/PixelgenTechnologies/pixelator/blob/482a2c7811de8a4ecc21cca521f9ce20dda3fbb0/src/pixelator/graph/backends/implementations/_networkx.py#L779

**Comment by ludvigla** [comment]
Perhaps we should throw a warning if this happens?

**Review by ambarrio — CHANGES_REQUESTED** [review]
I sent these ones also we were talking in the live review....

And I will check the tests.

File: src/pixelator/graph/backends/implementations/_networkx.py [hunk 1]
```diff
@@ -739,27 +739,43 @@ def subgraphs(self) -> Iterable[Graph]:
 
 
 def pmds_layout(
-    g: nx.Graph, pivots: int = 50, dim: int = 2, seed: Optional[int] = None
+    g: nx.Graph,
+    pivots: int = 50,
+    dim: int = 2,
+    weights: Optional[Union[np.ndarray, str]] = None,
+    add_noise: bool = False,
```
[the remark this answers is not in the document]
Comment (ludvigla) [reply src/pixelator/graph/backends/implementations/_networkx.py]: This option has been removed!

File: src/pixelator/graph/backends/implementations/_networkx.py [hunk 2]
```diff
@@ -818,3 +853,57 @@ def pmds_layout(
...
+    :rtype: np.array
+    :raises ValueError: if conditions are not met
+    """
+    # Check that k is an integer between 1 and 10
+    if not isinstance(k, int) and k < 1 or k > 10:
+        raise ValueError("'k' must be an integer between 1 and 10.")
+
+    A = nx.to_scipy_sparse_array(g, weight=None, nodelist=list(g.nodes), format="csr")
+
+    # Add 1 to the diagonal
```
[the remark this answers is not in the document]
Comment (ludvigla) [reply src/pixelator/graph/backends/implementations/_networkx.py]: Added comment

File: src/pixelator/graph/backends/implementations/_networkx.py [hunk 3]
```diff
@@ -818,3 +853,57 @@ def pmds_layout(
...
+    P = D @ A
+
+    # Compute the transition probabilities for a k-step walk
+    P_step = _mat_pow(P, k)
+
+    # Set diagonal to 0 and renormalize rows
+    P_step = P_step - sp.sparse.diags(P_step.diagonal(), format="csr")
+    P_step = P_step.multiply(1 / P_step.sum(axis=1))
+
+    P_step_bidirectional = P_step.multiply(P_step.T)
```
[the remark this answers is not in the document]
Comment (ludvigla) [reply src/pixelator/graph/backends/implementations/_networkx.py:910]: Added comment

**Review by ambarrio — COMMENTED** [review]
IS this resolving the test issues?

**Comment by ludvigla** [comment]
No it doesn't, but it improves computational speed. I need to fix the tests tomorrow.

**Review by ambarrio — APPROVED** [review]
I approve. Great work!

One question only, when we run this as default option, will we run it with "prob_dist" as default as weighting strategy?

**Comment by ludvigla** [comment]
> I approve. Great work!
> 
> One question only, when we run this as default option, will we run it with "prob_dist" as default as weighting strategy?

We have a test in place `test_generate_precomputed_layouts_for_components_with_all_components` to check that the default option is `wpmds_3d` which runs `pmds_layout` with `weights='prob_dist'`. I just quickly checked that the output makes sense with default settings, but I haven't actually tested the pipeline on real data. The default value when setting `--layout-algorithm` with the cli is `wpmds_3d` so it should be fine, but again, I haven't tested it on real data.

**Comment by ambarrio** [comment]
Great! We will test this now. And it makes sense with all the internal data you have shown. I think you should go ahead and merge.

**Pull request merged**